File: leetcode/exercises/3.1.7.hard.median_of_two_sorted_arrays.py
# -----------------------
# Created     : 06/12/2025
# Last Edited : 10/12/2025 
# Topics      : 
# Big O       :
# Problem Id  : 4. Median of Two Sorted Arrays
# References  : https://youtu.be/q6IEA26hvXc?si=_FV_HNLEili4HEvg
# Notes       : This solution broke my spirit
# -----------------------

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        A, B = nums1, nums2
        half = (len(nums1) + len(nums2)) // 2

        if len(B) < len(A):
            A, B = B, A

        l, r = 0, len(A)-1
        while True:
            m_A = l + (r - l) // 2
            m_B = half - m_A - 2

            A_left  = A[m_A]     if m_A >= 0 else float("-inf")
            A_right = A[m_A + 1] if (m_A + 1) < len(A) else float("inf")
            B_left  = B[m_B]     if m_B >= 0 else float("-inf")
            B_right = B[m_B + 1] if (m_B + 1) < len(B) else float("inf")

            if A_left <= B_right and B_left <= A_right:
                if (len(nums1) + len(nums2)) % 2 != 0:
                    return min(A_right, B_right)
                return (max(A_left, B_left) + min(A_right, B_right)) / 2
            elif A_left > B_right:
                r = m_A - 1
            else:
                l = m_A + 1

s = Solution()
print(s.findMedianSortedArrays([1,2],[3,4]))

File: leetcode/exercises/3.1.7.hard.median_of_two_sorted_arrays.js
// -----------------------
// Created     : 17/12/2025
// Last Edited : 18/12/2025 
// Topics      : 
// Big O       :
// Problem Id  : 4. Median of Two Sorted Arrays
// References  : https://youtu.be/q6IEA26hvXc?si=_FV_HNLEili4HEvg
// Notes       : This solution broke my spirit
// -----------------------

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var findMedianSortedArrays = function(nums1, nums2) {
    let a = nums1;
    let b = nums2;
    if (nums1.length > nums2.length) {
        a = nums2;
        b = nums1;
    }
    
    let half = Math.floor((a.length + b.length) / 2);

    let l = 0;
    let r = a.length - 1;
    while (true) {
        let m_1 = Math.floor(l + (r - l) / 2);
        let m_2 = (half - (m_1 + 1)) - 1;

        let amax = m_1 >= 0 ? a[m_1] : -Infinity; 
        let bmax = m_2 >= 0 ? b[m_2] : -Infinity; 

        let amin = m_1 + 1 < a.length ? a[m_1 + 1] : Infinity; 
        let bmin = m_2 + 1 < b.length ? b[m_2 + 1] : Infinity; 

        if ((amax <= bmin) && (bmax <= amin)) {
            if (((a.length + b.length) % 2) != 0 ) {
                return Math.min(amin, bmin);
            }
            return (Math.max(amax, bmax) + Math.min(amin, bmin)) / 2;
        } else if (amax > bmin) {
            r = m_1 - 1;
        } else {
            l = m_1 + 1;
        }
    }
    
};

console.log(findMedianSortedArrays([1,2],[3,4]));
